### PR TITLE
Object FIFO: Introduce new dynamic lowering

### PR DESCRIPTION
**Pull request by AndraBisca**
This PR introduces a new lowering for the `ObjectFifoAcquireOp`, `ObjectFifoSubviewAccessOp` and `ObjectFifoReleaseOp`. This lowering uses `IndexSwitchOps` to provide a way to rotate objectfifo accesses at runtime. This is combined with a global state buffer which is generated per core to keep track of the next elements to access for each objectfifo. The state is updated every time there is an `ObjectFifoReleaseOp` and is read before every switch.

A new flag is introduced to toggle the dynamic objectfifo lowering for all cores:
```
aiecc.py --dynamic-objFifos
```
To directly toggle for the lowering pass:
```
aie-opt --aie-objectFifo-stateful-transform=dynamic-objFifos %s
```

For finer control at the core level, a new optional attribute has been added to the `CoreOp` to toggle the dynamic objectfifo lowering:
```
OptionalAttr<BoolAttr>:$dynamic_objfifo_lowering
```

### CODE REVIEW

File: lib/Dialect/AIE/Transforms/AIEObjectFifoStatefulTransform.cpp [hunk 1]
```diff
@@ -967,6 +969,154 @@ struct AIEObjectFifoStatefulTransformPass
...
+    builder.create<memref::StoreOp>(size.getLoc(), newCounter, globalNextIndex,
+                                    ValueRange(ArrayRef({index.getResult()})));
+  }
+
+  // Function that generates the IR for objectfifo accesses to be handled at
+  // runtime.
+  LogicalResult dynamicGlobalObjectFifos(DeviceOp &device, OpBuilder &builder,
+                                         std::set<TileOp> objectFifoTiles) {
+    for (auto coreOp : device.getOps<CoreOp>()) {
+      if (objectFifoTiles.count(coreOp.getTileOp()) > 0) {
```
Comment (fifield) [inline lib/Dialect/AIE/Transforms/AIEObjectFifoStatefulTransform.cpp:994]: An early exit here can avoid indenting about one hundred lines of code:
```c++
if (objectFifoTiles.count(coreOp.getTileOp()) <= 0)
  continue;
```

File: programming_examples/dynamic_object_fifo/nested_loops/test.cpp [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+  while (std::getline(instr_file, line)) {
+    std::istringstream iss(line);
+    uint32_t a;
+    if (!(iss >> std::hex >> a)) {
+      throw std::runtime_error("Unable to parse instruction file\n");
+    }
+    instr_v.push_back(a);
+  }
+  return instr_v;
+}
```
Comment (fifield) [inline programming_examples/dynamic_object_fifo/nested_loops/test.cpp:48]:
```suggestion
#include "test_utils.h"
```

File: programming_examples/dynamic_object_fifo/nested_loops/test.cpp [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+
+int main(int argc, const char *argv[]) {
+
+  std::vector<uint32_t> instr_v = load_instr_sequence(INSTS_TXT);
```
Comment (fifield) [inline programming_examples/dynamic_object_fifo/nested_loops/test.cpp:52]:
```suggestion
  std::vector<uint32_t> instr_v = test_utils::load_instr_sequence(INSTS_TXT);
```

File: programming_examples/dynamic_object_fifo/ping_pong/test.cpp [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+  while (std::getline(instr_file, line)) {
+    std::istringstream iss(line);
+    uint32_t a;
+    if (!(iss >> std::hex >> a)) {
+      throw std::runtime_error("Unable to parse instruction file\n");
+    }
+    instr_v.push_back(a);
+  }
+  return instr_v;
+}
```
Comment (fifield) [inline programming_examples/dynamic_object_fifo/ping_pong/test.cpp:44]:
```suggestion
#include "test_utils.h"
```

File: programming_examples/dynamic_object_fifo/ping_pong/test.cpp [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+
+int main(int argc, const char *argv[]) {
+
+  std::vector<uint32_t> instr_v = load_instr_sequence(INSTS_TXT);
```
Comment (fifield) [inline programming_examples/dynamic_object_fifo/ping_pong/test.cpp:48]:
```suggestion
  std::vector<uint32_t> instr_v = test_utils::load_instr_sequence(INSTS_TXT);
```

File: programming_examples/dynamic_object_fifo/two_core_sliding_window/test.cpp [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+      throw std::runtime_error("Unable to parse instruction file\n");
+    }
+    instr_v.push_back(a);
+  }
+  return instr_v;
+}
+
+int main(int argc, const char *argv[]) {
+
+  std::vector<uint32_t> instr_v = load_instr_sequence(INSTS_TXT);
```
Comment (fifield) [inline programming_examples/dynamic_object_fifo/two_core_sliding_window/test.cpp:51]:
```suggestion
  std::vector<uint32_t> instr_v = test_utils::load_instr_sequence(INSTS_TXT);
```

File: programming_examples/dynamic_object_fifo/reduction/test.cpp [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+      throw std::runtime_error("Unable to parse instruction file\n");
+    }
+    instr_v.push_back(a);
+  }
+  return instr_v;
+}
+
+int main(int argc, const char *argv[]) {
+
+  std::vector<uint32_t> instr_v = load_instr_sequence(INSTS_TXT);
```
Comment (fifield) [inline programming_examples/dynamic_object_fifo/reduction/test.cpp:51]:
```suggestion
  std::vector<uint32_t> instr_v = test_utils::load_instr_sequence(INSTS_TXT);
```

File: programming_examples/dynamic_object_fifo/sliding_window/test.cpp [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+      throw std::runtime_error("Unable to parse instruction file\n");
+    }
+    instr_v.push_back(a);
+  }
+  return instr_v;
+}
+
+int main(int argc, const char *argv[]) {
+
+  std::vector<uint32_t> instr_v = load_instr_sequence(INSTS_TXT);
```
Comment (fifield) [inline programming_examples/dynamic_object_fifo/sliding_window/test.cpp:51]:
```suggestion
  std::vector<uint32_t> instr_v = test_utils::load_instr_sequence(INSTS_TXT);
```

**Review by fifield — CHANGES_REQUESTED** [review]
The new programming examples need lit scripts. Otherwise they will never be tested.

File: lib/Dialect/AIE/Transforms/AIEObjectFifoStatefulTransform.cpp (hunk 1 above)
Comment (AndraBisca) [reply lib/Dialect/AIE/Transforms/AIEObjectFifoStatefulTransform.cpp:994]: Good point! Updated.

**Review by fifield — APPROVED** [review]

**Pull request merged**